command/src/access_logs.rs

```diff
@@ -0,0 +1,207 @@
+use std::{collections::BTreeMap, mem::ManuallyDrop, net::SocketAddr};
+
+use rusty_ulid::Ulid;
+use time::Duration;
+
+use crate::proto::command::{
+    BinaryAccessLog, BinaryEndpoint, HttpMethod, RequestUlid, TcpEndpoint,
+};
+
+/// This uses unsafe to creates a "fake" owner of the underlying data.
+/// Beware that for the compiler it is as legitimate as the original owner.
+/// So you have to elide one of them (with std::mem::forget or ManuallyDrop)
+/// before it is drop to avoid a double free.
+///
+/// This trait works on &T and Option<&T> types
+trait DuplicateOwnership {
+    type Target;
+    /// Don't forget to use std::mem::forget or ManuallyDrop over one of your owners
+    unsafe fn duplicate(self) -> Self::Target;
+}
+
+impl<T> DuplicateOwnership for &T {
+    type Target = T;
+    unsafe fn duplicate(self) -> T {
+        std::ptr::read(self as *const T)
+    }
+}
+impl<'a, T> DuplicateOwnership for Option<&'a T>
+where
+    T: ?Sized,
+    &'a T: DuplicateOwnership + 'a,
+{
+    type Target = Option<<&'a T as DuplicateOwnership>::Target>;
+    unsafe fn duplicate(self) -> Self::Target {
+        self.map(|t| t.duplicate())
+    }
+}
+impl DuplicateOwnership for &str {
+    type Target = String;
+    unsafe fn duplicate(self) -> Self::Target {
+        String::from_raw_parts(self.as_ptr() as *mut _, self.len(), self.len())
+    }
+}
+impl<T> DuplicateOwnership for &[T] {
+    type Target = Vec<T>;
+    unsafe fn duplicate(self) -> Self::Target {
+        Vec::from_raw_parts(self.as_ptr() as *mut _, self.len(), self.len())
+    }
+}
+
+pub fn prepare_user_agent(user_agent: &str) -> String {
+    let mut user_agent = user_agent.replace(' ', "_");
+    let mut ua_bytes = std::mem::take(&mut user_agent).into_bytes();
+    if let Some(last) = ua_bytes.last_mut() {
+        if *last == b',' {
+            *last = b'!'
+        }
+    }
+    unsafe { String::from_utf8_unchecked(ua_bytes) }
+}
+
+pub struct LogDuration(pub Option<Duration>);
+
+#[derive(Debug)]
+pub struct LogContext<'a> {
+    pub request_id: Ulid,
+    pub cluster_id: Option<&'a str>,
+    pub backend_id: Option<&'a str>,
+}
+
+#[derive(Clone, Copy)]
+pub enum MethodRecord {
+    Get,
+    Post,
+    Head,
+    Options,
+    Put,
+    Delete,
+    Trace,
+    Connect,
+    Other,
```

Comment:
In kubernetes, there are also Watch, List and ListAndWatch. 
Their are not standard, but it will be great to add them with a reference to their implementation/documentation as we want to provide an ingress controller and/or an api gateway using the kubernetes cloud controller manager component.

Reply:
it is a string now :+1: